Tugas 1.1 - REGRESI LINIER
----------------------------
Nama : Astri Cahyaningtyas

Kelas : IF-39-02

NIM : 1301154548

In [167]:
import numpy as np
np.set_printoptions(precision=4, suppress=True)
import pandas as pd

Load Data Training Set

In [117]:
def load_data(filename):
    dtrain = pd.read_csv(filename, delimiter =',').as_matrix()
    X = dtrain[:,0:1]
    Y = dtrain[:,-1:]
    return X,Y

X_train_1, Y_train_1 = load_data('training_set_1.csv')

#print(X_train_1)
#print(Y_train_1)

Question 1
=====

Jika dibangun model regresi $y_i=f(x_i )=w_0+w_1 x_i+w_2 x_i^2+w_3 x_i^3$
maka dihasilkan prakiraan weight vector $w = (w_0, w_1, w_2, w_3) = (1.089, 0.840, 0.943, 3.111)$

Bangun Matrix X berdasarkan model regresi

$y_i=f(x_i )=w_0+w_1 x_i+w_2 x_i^2+w_3 x_i^3$

In [169]:
def model_regresi_1(X_train_1):
    X_ones = np.ones((X_train_1.shape[0],1))
    return np.hstack((X_ones , X_train_1 , X_train_1**2 , X_train_1**3)) 
    
X_matrix_1 = model_regresi_1(X_train_1)
X_matrix_1

array([[  1.    ,  -2.907 ,   8.4506, -24.566 ],
       [  1.    ,  -2.745 ,   7.535 , -20.6836],
       [  1.    ,  -2.637 ,   6.9538, -18.3371],
       [  1.    ,  -2.42  ,   5.8564, -14.1725],
       [  1.    ,  -2.276 ,   5.1802, -11.7901],
       [  1.    ,  -2.101 ,   4.4142,  -9.2742],
       [  1.    ,  -1.973 ,   3.8927,  -7.6804],
       [  1.    ,  -1.857 ,   3.4484,  -6.4038],
       [  1.    ,  -1.643 ,   2.6994,  -4.4352],
       [  1.    ,  -1.408 ,   1.9825,  -2.7913],
       [  1.    ,  -1.248 ,   1.5575,  -1.9438],
       [  1.    ,  -0.963 ,   0.9274,  -0.8931],
       [  1.    ,  -0.766 ,   0.5868,  -0.4495],
       [  1.    ,  -0.654 ,   0.4277,  -0.2797],
       [  1.    ,  -0.494 ,   0.244 ,  -0.1206],
       [  1.    ,  -0.43  ,   0.1849,  -0.0795],
       [  1.    ,  -0.277 ,   0.0767,  -0.0213],
       [  1.    ,  -0.108 ,   0.0117,  -0.0013],
       [  1.    ,   0.112 ,   0.0125,   0.0014],
       [  1.    ,   0.235 ,   0.0552,   0.013 ],
       [  1.    ,   

Normal Equation

$w=(x^T  x)^(-1) x^T y $


In [119]:
def normal_equation(X_matrix_1,Y_train_1):
    Xt = X_matrix_1.conj().transpose()
    return np.linalg.pinv(Xt.dot(X_matrix_1)).dot(Xt).dot(Y_train_1)

w_1 = normal_equation(X_matrix_1, Y_train_1)
w_1

array([[ 1.0894],
       [-0.8402],
       [-0.9432],
       [-3.1117]])

Answer 1
====

B

Question 2
=====

Salah satu model regresi multivariate linear adalah $y_i = f(x_i) = (0.4x_i + 3,1)$, dan jika digunakan
maka SSEtrain = 43454,587.

Answer 2
====

S

Question 3
=====

Regresi multivariate non-linear dan regresi univariate linear keduanya lebih baik daripada regresi
univariate non-linear jika digunakan untuk pemodelan training set 1.

Answer3
====

S

Question 4
=====

Jika menggunakan model regresi $y_i = f(x_i) = 21.274 - 10.608x_i + 0.921x_2i$
untuk 5 record/data tes (xi, yi) secara berurutan antara lain (x1, y1) = $(-2.864, 73.128), (x2, y2) = (1.489, 20.271), (x3, y3) = (-0.235,-3.220), (x4, y4) = (1.092, 0.347),$ dan $(x5, y5) = (2.898,-75.691)$ maka didapatkan SSEtes = 8686.216.

In [127]:
y_1 = 73.128
y_2 = 20.271
y_3 = -3.220
y_4 = 0.347
y_5 = -75.691

y_test_1 = 21.274 - (10.608 * (-2.864)) + (0.921 * (-2.864**2))
y_test_2 = 21.274 - (10.608 * 1.489) + (0.921 * (1.489**2))
y_test_3 = 21.274 - (10.608 * (-0.235)) + (0.921 * (-0.235**2))
y_test_4 = 21.274 - (10.608 * 1.092) + (0.921 * (-0.235**2))
y_test_5 = 21.274 - (10.608 * 2.898) + (0.921 * (2.898**2))

sse_5 = ((y_test_1-y_1)**2)+((y_test_2-y_2)**2)+((y_test_3-y_3)**2)+((y_test_4-y_4)**2)+((y_test_5-y_5)**2)
sse_5

7286.820691689777

Answer 4
====

S

Question 5
=====

Jika dibangun model regresi $y_i = f(x_i) = w_xi + a$ maka $w = -18.757$ dan $a = -2.745$

In [123]:
def model_regresi_5(X_train_1):
    X_ones = np.ones((X_train_1.shape[0],1))
    return np.hstack((X_train_1, X_ones)) 
    
X_matrix_5 = model_regresi_5(X_train_1)
X_matrix_5

array([[-2.907,  1.   ],
       [-2.745,  1.   ],
       [-2.637,  1.   ],
       [-2.42 ,  1.   ],
       [-2.276,  1.   ],
       [-2.101,  1.   ],
       [-1.973,  1.   ],
       [-1.857,  1.   ],
       [-1.643,  1.   ],
       [-1.408,  1.   ],
       [-1.248,  1.   ],
       [-0.963,  1.   ],
       [-0.766,  1.   ],
       [-0.654,  1.   ],
       [-0.494,  1.   ],
       [-0.43 ,  1.   ],
       [-0.277,  1.   ],
       [-0.108,  1.   ],
       [ 0.112,  1.   ],
       [ 0.235,  1.   ],
       [ 0.498,  1.   ],
       [ 0.566,  1.   ],
       [ 0.869,  1.   ],
       [ 0.958,  1.   ],
       [ 1.015,  1.   ],
       [ 1.207,  1.   ],
       [ 1.427,  1.   ],
       [ 1.905,  1.   ],
       [ 1.946,  1.   ],
       [ 2.133,  1.   ],
       [ 2.438,  1.   ],
       [ 2.71 ,  1.   ],
       [ 2.896,  1.   ],
       [ 2.994,  1.   ]])

In [124]:
def normal_equation_5(X_matrix_5,Y_train_1):
    Xt = X_matrix_5.conj().transpose()
    return np.linalg.pinv(Xt.dot(X_matrix_5)).dot(Xt).dot(Y_train_1)

w_5 = normal_equation(X_matrix_5, Y_train_1)
w_5

array([[-18.7566],
       [ -2.745 ]])

Answer 5
====

B

Question 6
=====

Jika dibangun model regresi $y_i = f(x_i) = w_0 + w_1 x_i + w_2 x_i^2$ maka dihasilkan prakiraan weight vector $w = (w_0,w_1,w_2) = (1.590,-17.824,-4.118)$

In [120]:
def model_regresi_6(X_train_1):
    X_ones = np.ones((X_train_1.shape[0],1))
    return np.hstack((X_ones , X_train_1 , X_train_1**2)) 
    
X_matrix_6 = model_regresi_6(X_train_1)
X_matrix_6

array([[ 1.    , -2.907 ,  8.4506],
       [ 1.    , -2.745 ,  7.535 ],
       [ 1.    , -2.637 ,  6.9538],
       [ 1.    , -2.42  ,  5.8564],
       [ 1.    , -2.276 ,  5.1802],
       [ 1.    , -2.101 ,  4.4142],
       [ 1.    , -1.973 ,  3.8927],
       [ 1.    , -1.857 ,  3.4484],
       [ 1.    , -1.643 ,  2.6994],
       [ 1.    , -1.408 ,  1.9825],
       [ 1.    , -1.248 ,  1.5575],
       [ 1.    , -0.963 ,  0.9274],
       [ 1.    , -0.766 ,  0.5868],
       [ 1.    , -0.654 ,  0.4277],
       [ 1.    , -0.494 ,  0.244 ],
       [ 1.    , -0.43  ,  0.1849],
       [ 1.    , -0.277 ,  0.0767],
       [ 1.    , -0.108 ,  0.0117],
       [ 1.    ,  0.112 ,  0.0125],
       [ 1.    ,  0.235 ,  0.0552],
       [ 1.    ,  0.498 ,  0.248 ],
       [ 1.    ,  0.566 ,  0.3204],
       [ 1.    ,  0.869 ,  0.7552],
       [ 1.    ,  0.958 ,  0.9178],
       [ 1.    ,  1.015 ,  1.0302],
       [ 1.    ,  1.207 ,  1.4568],
       [ 1.    ,  1.427 ,  2.0363],
       [ 1.    ,  1.905 ,  3

In [121]:
def normal_equation_6(X_matrix_6,Y_train_1):
    Xt = X_matrix_6.conj().transpose()
    return np.linalg.pinv(Xt.dot(X_matrix_6)).dot(Xt).dot(Y_train_1)

w_6 = normal_equation_6(X_matrix_6, Y_train_1)
w_6

array([[  1.59  ],
       [-18.725 ],
       [ -1.4181]])

Answer 6
====

S

Question 7
=====

Regresi multivariate linear dan regresi univariate non-linear keduanya lebih baik daripada regresi
multivariate non-linear jika digunakan untuk pemodelan training set 2.

Answer 7
====

B

In [137]:
def load_data_2(filename):
    dtrain = pd.read_csv(filename, delimiter =',').as_matrix()
    X1 = dtrain[:,0:1]
    X2 = dtrain[:,1:2]
    Y = dtrain[:,-1:]
    return X1, X2,Y

X_train_2_1, X_train_2_2, Y_train_2 = load_data_2('training_set_2.csv')

#print(X_train_2_1)
#print(X_train_2_2)
#print(Y_train_2)

Question 8
=====

Jika dibangun model regresi $y_i = f(x_i1, x_i2) = w_0+w_1x_i1+w_2x_i2+w_3(x_i1x_i2)+w4(x_i1+x_i2)3$ maka
dihasilkan prakiraan weight vector $w = (w_0, w_1, w_2, w_3, w_4) = (4.514, -1.678, 1.842, -0.002, -0.005)$

In [159]:
def model_regresi_8(X_train_2_1, X_train_2_2):
    X_ones = np.ones((X_train_2_1.shape[0],1)) 
    return np.hstack((X_ones , X_train_2_1 , X_train_2_2 , (X_train_2_1 * X_train_2_2), (X_train_2_1 + X_train_2_2)**3)) 

X_matrix_8 = model_regresi_8(X_train_2_1, X_train_2_2)
X_matrix_8

array([[     1.    ,    -14.741 ,    -14.99  ,    220.9676, -26280.193 ],
       [     1.    ,    -13.909 ,    -13.588 ,    188.9955, -20790.0695],
       [     1.    ,    -13.511 ,    -12.161 ,    164.3073, -16919.1723],
       [     1.    ,    -11.676 ,    -11.112 ,    129.7437, -11833.6476],
       [     1.    ,    -11.182 ,     -9.973 ,    111.5181,  -9467.5823],
       [     1.    ,    -10.355 ,     -8.59  ,     88.9494,  -6799.6073],
       [     1.    ,    -10.153 ,     -8.248 ,     83.7419,  -6230.5197],
       [     1.    ,     -8.75  ,     -7.99  ,     69.9125,  -4691.01  ],
       [     1.    ,     -8.033 ,     -7.722 ,     62.0308,  -3910.7065],
       [     1.    ,     -6.502 ,     -6.199 ,     40.3059,  -2048.8669],
       [     1.    ,     -4.612 ,     -5.304 ,     24.462 ,   -975.0111],
       [     1.    ,     -3.93  ,     -5.129 ,     20.157 ,   -743.4312],
       [     1.    ,     -2.946 ,     -4.6   ,     13.5516,   -429.6852],
       [     1.    ,     -2.768 ,     

In [160]:
def normal_equation_8(X_matrix_8,Y_train_2):
    Xt = X_matrix_8.conj().transpose()
    return np.linalg.pinv(Xt.dot(X_matrix_8)).dot(Xt).dot(Y_train_2)

w_8 = normal_equation_8(X_matrix_8, Y_train_2)
w_8

array([[ 4.5143],
       [-1.6783],
       [ 1.8422],
       [-0.0023],
       [-0.0048]])

Answer 8
====

B

Question 9
=====

Jika menggunakan model regresi $y_i = f(x_i1, x_i2) = 4.443-8.112x_i1 +3.028x_i2 -0.025x_i1 x_i2$ untuk input (x11, x12) = (-7.767,-7.111) dan (x21, x22) = (-0.333, 0.857) maka masing-masing diperoleh $y_1 = 44.536$ dan $y_2 = 9.746$

In [166]:
y_test_1 = 4.443 - (8.112 * (-7.767)) + (3.028 * (-7.111)) - (0.025 * ((-7.767) * (-7.111)))
y_test_2 = 4.443 - (8.112 * (-0.333)) + (3.028 * (0.857)) - (0.025 * ((-0.333) * (0.857)))
                                         
print(y_test_1)
print(y_test_2)

44.536017574999995
9.746426524999999


Answer 9
====

B

Question 10
======

Jika dibangun model regresi $y_i = f(x_i1, x_i2) = w_0 + w_1 x_i^3 + w_2x_i^3$ maka dihasilkan prakiraan weight vector $w = (w_0,w_1,w_2) = (4.373,-0.022,-0.015)$

In [162]:
def model_regresi_10(X_train_2_1, X_train_2_2):
    X_ones = np.ones((X_train_2_1.shape[0],1)) 
    return np.hstack((X_ones , (X_train_2_1**3) , (X_train_2_2**3))) 

X_matrix_10 = model_regresi_10(X_train_2_1, X_train_2_2)
X_matrix_10

array([[    1.    , -3203.1763, -3368.2545],
       [    1.    , -2690.839 , -2508.8033],
       [    1.    , -2466.3942, -1798.4893],
       [    1.    , -1591.7771, -1372.0714],
       [    1.    , -1398.1651,  -991.9219],
       [    1.    , -1110.3255,  -633.8398],
       [    1.    , -1046.6059,  -561.1073],
       [    1.    ,  -669.9219,  -510.0824],
       [    1.    ,  -518.3622,  -460.4573],
       [    1.    ,  -274.8786,  -238.2127],
       [    1.    ,   -98.0997,  -149.2143],
       [    1.    ,   -60.6985,  -134.9268],
       [    1.    ,   -25.5681,   -97.336 ],
       [    1.    ,   -21.2079,   -79.7847],
       [    1.    ,    -6.8915,   -28.7933],
       [    1.    ,    -2.4658,    -6.4452],
       [    1.    ,    -0.6745,     0.3299],
       [    1.    ,     1.3456,     1.8428],
       [    1.    ,    15.1793,     3.9219],
       [    1.    ,    32.1271,    21.485 ],
       [    1.    ,    36.7931,    49.9987],
       [    1.    ,    52.65  ,   193.6021],
       [  

In [163]:
def normal_equation_10(X_matrix_10,Y_train_2):
    Xt = X_matrix_10.conj().transpose()
    return np.linalg.pinv(Xt.dot(X_matrix_10)).dot(Xt).dot(Y_train_2)

w_10 = normal_equation_10(X_matrix_10, Y_train_2)
w_10

array([[ 4.3732],
       [-0.022 ],
       [-0.0152]])

Answer 10
=====

B

Semua Jawaban
=======

1 B

2 S

3 S

4 S

5 B

6 S

7 B

8 B

9 B

10 B